src/compute/src/sink/subscribe.rs

```diff
@@ -152,45 +153,51 @@ fn subscribe<G>(
 /// otherwise sends an indication that the protocol has finished without completion.
 struct SubscribeProtocol {
     pub sink_id: GlobalId,
+    pub sink_as_of: Antichain<Timestamp>,
     pub subscribe_response_buffer: Option<Rc<RefCell<Vec<(GlobalId, SubscribeResponse)>>>>,
     pub prev_upper: Antichain<Timestamp>,
 }
 
 impl SubscribeProtocol {
     fn send_batch(&mut self, upper: Antichain<Timestamp>, rows: &mut Vec<(Timestamp, Row, Diff)>) {
-        if self.prev_upper != upper {
-            let mut ship = Vec::new();
-            let mut keep = Vec::new();
-            differential_dataflow::consolidation::consolidate_updates(rows);
-            for (time, data, diff) in rows.drain(..) {
-                if upper.less_equal(&time) {
-                    keep.push((time, data, diff));
-                } else {
-                    ship.push((time, data, diff));
-                }
-            }
-            *rows = keep;
+        // Only send a batch if both conditions hold:
+        //  a) `upper` has reached or passed the sink's `as_of` frontier.
+        //  b) `upper` is different from when we last sent a batch.
+        if !PartialOrder::less_equal(&self.sink_as_of, &upper) || upper == self.prev_upper {
```

Comment:
> I might be missing something, but when would we expect that we get a call to this function with `PartialOrder::less_equal(&self.sink_as_of, &upper)`?

I assume you meant to write `!PartialOrder::less_equal(&self.sink_as_of, &upper)` (i.e. the upper has not yet advanced to the sink's as_of)? This can happen at the beginning of a subscribe. The `should_emit_as_of` filters data updates, but it does not filter frontier updates. Note that `send_batch` gets called every time the subscribe closure is invoked, regardless of whether there is any data or not. After all, subscribe batches also communicate frontier updates.

> My concern is that we are now explicitly ignoring some information that was previously sent for a `SUBSCRIBE`.

Yes, we are ignoring frontier updates before the `as_of`, which is exactly the intention. We are not ignoring data updates, for two reasons:
* As you point out, data updates are already filtered against the `as_of` in the subscribe operator, so no updates before the `as_of` should get here.
* Even if updates before the `as_of` got here somehow, we are not throwing them away. We simply return but keep the `rows` buffer untouched. If you look at the subscribe operator, you see that it re-uses the buffer and submits its contents again the next time that it wants to send a batch. Only if we drain the `rows` buffer do we actually remove the data from the subscribe operators responsibility.

> I am just a bit confused about what the additional filtering here is achieving.

I think I answered that above!

Reply:
Maybe this would be clearer if we did the frontier filtering in the subscribe operator instead? I added it to `send_batch` because this is where the `SubscribeResponse` is constructed, so it is easy to reason about the correctness of its contents locally there.

Reply:
Ah, I see, your explanation makes a lot of sense to me, thanks! I think that I got confused about `send_batch` not taking over the given `rows` buffer sometimes and letting it sit back in the operator (named as `results` there, as opposed to `rows`, which in the operator names the input batch). It's kind of `maybe_send_batch_maybe_leave_the_rows_buffer_alone_to_the_next_time_you_call_me`. I'd find it a clearer formulation if `send_batch` actually sent the batch and all the conditions for a sending were checked outside. Then, you'd know that `results` would get flushed by `send_batch`. But that is just a personal preference and the refactor might exceed the scope of this PR. So no opposition from my side to keeping it as is!

Reply:
I kind of like the fact that `send_batch` can ensure that batches is sends are actually valid, rather than relying on the caller to provide valid batches. That said, I understand the confusion. Maybe we could have `send_batch` only assert things and then rely on the caller. But yes, I would consider that out-of-scope here. Instead I'll add a docstring describing the `send_batch` contract.